src/core/useSpriteLoader.tsx

```diff
@@ -26,355 +75,409 @@ export const checkIfFrameIsEmpty = (frameData: Uint8ClampedArray) => {
+      } else {
+        return { rows: 0, columns: 0, frameWidth: 0, frameHeight: 0, emptyFrames: [] }
       }
+    },
+    [canvasRenderingContext2DSettings]
+  )
 
-      //scale ratio for stadalone sprite
-      spriteDataRef.current.frames = calculateScaleRatio(spriteDataRef.current.frames)
-    } else if (_spriteTexture) {
-      spriteDataRef.current = json
-      spriteDataRef.current.frames = parseFrames()
-
-      totalFrames.current = Array.isArray(json.frames) ? json.frames.length : Object.keys(json.frames).length
-      const { w, h } = getFirstItem(json.frames).sourceSize
-      aspect = calculateAspectRatio(w, h, aspectFactor, v)
-    }
+  // calculate scale ratio for the frames
+  const calculateScaleRatio = React.useCallback((frames: FrameData[] | Record<string, FrameData[]>) => {
+    // Helper function to calculate scale ratio for an array of frames
+    const processFrameArray = (frameArray: FrameData[]) => {
+      // Find the largest frame
+      let largestFrame: { w: number; h: number; area: number } | null = null
+
+      for (const frame of frameArray) {
+        const { w, h } = frame.frame
+        const area = w * h
+        if (!largestFrame || area > largestFrame.area) {
+          largestFrame = { w, h, area }
+        }
+      }
 
-    setSpriteData(spriteDataRef.current)
-    // @ts-ignore
-    if ('encoding' in _spriteTexture) _spriteTexture.encoding = 3001 // sRGBEncoding
-    // @ts-ignore
-    else _spriteTexture.colorSpace = THREE.SRGBColorSpace
-    setSpriteTexture(_spriteTexture)
-    setSpriteObj({
-      spriteTexture: _spriteTexture,
-      spriteData: spriteDataRef.current,
-      aspect: aspect,
-    })
-  }
+      // Set scaleRatio property on each frame
+      return frameArray.map((frame) => {
+        const { w, h } = frame.frame
+        const area = w * h
+        const scaleRatio = largestFrame ? (area === largestFrame.area ? 1 : Math.sqrt(area / largestFrame.area)) : 1
 
-  const getRowsAndColumns = (texture: THREE.Texture, totalFrames: number) => {
-    if (texture.image) {
-      const canvas = document.createElement('canvas')
-      const ctx = canvas.getContext('2d', canvasRenderingContext2DSettings)
+        return {
+          ...frame,
+          scaleRatio,
+        }
+      })
+    }
 
-      if (!ctx) {
-        throw new Error('Failed to get 2d context')
+    // Handle both array and record cases
+    if (Array.isArray(frames)) {
+      return processFrameArray(frames)
+    } else {
+      // Process each animation sequence separately
+      const result: Record<string, any[]> = {}
+      for (const key in frames) {
+        result[key] = processFrameArray(frames[key])
       }
+      return result
+    }
+  }, [])
 
-      canvas.width = texture.image.width
-      canvas.height = texture.image.height
+  // for frame based JSON Hash sprite data
+  const parseFrames = React.useCallback((): any => {
+    const sprites: Record<string, any> = {}

```

Comment:
also here

Reply:
done, it needed some typescript gymnastics, but it's more streamlined now